ssl/priority_queue.c

```diff
@@ -0,0 +1,390 @@
+    while (n > 0) {
+        const size_t p = (n - 1) / 2;
+
+        ASSERT_USED(pq, p);
+        pqueue_swap_elem(pq, n, p);
+        n = p;
+    }
+}
+
+/*
+ * Move an element down to its correct position to restore the partial
+ * order pre-condition.
+ */
+static ossl_inline void pqueue_move_down(OSSL_PQUEUE *pq, size_t n)
+{
+    struct pq_heap_s *h = pq->heap;
+
+    ASSERT_USED(pq, n);
+    while (n > 0) {
+        const size_t p = (n - 1) / 2;
+
+        ASSERT_USED(pq, p);
+        if (pq->compare(h[n].data, h[p].data) >= 0)
+            break;
+        pqueue_swap_elem(pq, n, p);
+        n = p;
+    }
+}
+
+/*
+ * Move an element up to its correct position to restore the partial
+ * order pre-condition.
+ */
+static ossl_inline void pqueue_move_up(OSSL_PQUEUE *pq, size_t n)
+{
+    struct pq_heap_s *h = pq->heap;
+    size_t p = n * 2 + 1;
+
+    ASSERT_USED(pq, n);
+    if (pq->htop > p + 1) {
+        ASSERT_USED(pq, p);
+        ASSERT_USED(pq, p + 1);
+        if (pq->compare(h[p].data, h[p + 1].data) > 0)
+            p++;
+    }
+    while (pq->htop > p && pq->compare(h[p].data, h[n].data) < 0) {
+        ASSERT_USED(pq, p);
+        pqueue_swap_elem(pq, n, p);
+        n = p;
+        p = n * 2 + 1;
+        if (pq->htop > p + 1) {
+            ASSERT_USED(pq, p + 1);
+            if (pq->compare(h[p].data, h[p + 1].data) > 0)
+                p++;
+        }
+    }
+}
+
+int ossl_pqueue_push(OSSL_PQUEUE *pq, void *data, size_t *elem)
+{
+    size_t n, m;
+
+    if (!ossl_pqueue_reserve(pq, 1))
+        return 0;
+
+    n = pq->htop++;
+    m = pq->freelist;
+    pq->freelist = pq->elements[m].posn;
+
+    pq->heap[n].data = data;
+    pq->heap[n].index = m;
+
+    pq->elements[m].posn = n;
+#ifndef NDEBUG
+    pq->elements[m].used = 1;
+#endif
+    pqueue_move_down(pq, n);
+    if (elem != NULL)
+        *elem = m;
+    return 1;
+}
+
+void *ossl_pqueue_peek(const OSSL_PQUEUE *pq)
+{
+    if (pq->htop > 0) {
+        ASSERT_USED(pq, 0);
+        return pq->heap->data;
+    }
+    return NULL;
+}
+
+void *ossl_pqueue_pop(OSSL_PQUEUE *pq)
+{
+    void *res;
+    size_t elem;
+
+    if (pq == NULL) {
+        ERR_raise(ERR_LIB_SSL, CRYPTO_R_INVALID_NULL_ARGUMENT);
+        return NULL;
+    }
+    if (pq->htop == 0) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_INTERNAL_ERROR);
+        return NULL;
+    }
+
+    ASSERT_USED(pq, 0);
+    res = pq->heap->data;
+    elem = pq->heap->index;
+
+    if (--pq->htop != 0) {
+        pqueue_move_elem(pq, pq->htop, 0);
+        pqueue_move_up(pq, 0);
+    }
+
+    pq->elements[elem].posn = pq->freelist;
+    pq->freelist = elem;
+#ifndef NDEBUG
+    pq->elements[elem].used = 0;
+#endif
+    return res;
+}
+
+void *ossl_pqueue_remove(OSSL_PQUEUE *pq, size_t elem)
+{
+    size_t n;
+
+    if (pq == NULL) {
+        ERR_raise(ERR_LIB_SSL, CRYPTO_R_INVALID_NULL_ARGUMENT);
+        return NULL;
+    }
+    if (elem >= pq->hmax) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_INTERNAL_ERROR);
```

Comment:
Isn't this really a caller error...so I'm wondering if internal error is the right reason code here. Since this is an internal API, I guess maybe its ok??

Reply:
Not sure what's best.  Not raising any errors would be the alternative: the caller should check and catch the NULL return & raise something more interesting.

Reply:
`ERR_R_PASSED_INVALID_ARGUMENT` seems better